### ISSUE TITLE
Include new Chocolatey package for Windows in installation list 🚀

### ISSUE THREAD

**Issue by nzbart** [dandavison/delta#197]
I have created a package for Windows using the Chocolatey package manager. It can be installed via:

```powershell
choco install delta
```

I now realise that I probably should have called the package `git-delta` to match the existing packages. Do you think this is important enough to change? Or should I leave the package name matching the executable name?

The package page is at: https://chocolatey.org/packages/delta

**Comment by dandavison** [comment]
Hi @nzbart, thanks a lot for doing this. I think the name is great: I don't see much confusion arising from the fact that the MacOS/Linux package managers use git-delta as the package name. Everyone's going to have to deal with the fact that the executable's name is "delta", and as you say, that will only be easier for chocolatey users.

So, thanks again for doing it. May I ask you -- do you think it would be helpful for the Chocolatey package page to mention either of the following issues:

- terminal emulators that work well with delta (especially, support 24 bit colors)
- the version of less.exe to use and how to install

The latter is discussed in the [delta README](https://github.com/dandavison/delta#windows), but the former not yet.

The considerations regarding less.exe are very similar (or identical) to those regarding installing a less.exe that works well with bat. I notice that the bat chocolatey package actually bundles less.exe to ensure users have a working combination: https://chocolatey.org/packages/Bat#dependencies So, possibly that's something the delta chocolatey package could aim for in the future?

**Comment by nzbart** [comment]
Hi @dandavison, I'm happy to put any information you think might be helpful on the page. Regarding your two points specifically:

**24 bit colour support**
I haven't personally had any problems with colour support, and as far as I can tell, the two major Microsoft supported terminals both support 24 bit colour:
- [The legacy Windows console](https://devblogs.microsoft.com/commandline/24-bit-color-in-the-windows-console/)
- The new Windows Terminal

I personally use [kitty](https://sw.kovidgoyal.net/kitty/) as my terminal under WSL, so I get a Linux-like experience anyway. However, I have tested it briefly in both Microsoft terminals, and it works just as well.

**Problems with less pager**
I also haven't noticed any problems with the version of less.exe that ships with Git. Do you know specifically what issues I'd be looking for? I had a look at the documentation for Bat, and they don't seem to indicate any problems beyond less.exe being missing. Their package also depends directly on a standard less.exe Chocolatey package that I couldn't see had any special patches. I'm a bit reluctant to overwrite the user's existing version of less.exe if it's not required.

Note that it's quite likely that some of the issues that people ran into originally have since been resolved in the Windows Git version of less.exe.

Basically, I'm saying that delta works great out of the box in Windows in my experience!

**Comment by dandavison** [comment]
Thanks @nzbart. I've added Chocolatey to the README in 2227adfb15a7b714d0d33afe25ac7ecd08f171f2.

What's your less version?  @lzybkr has recently fixed some bugs in less that were surfacing when using delta. My understanding is that not all his fixes are in `gwsw/less`, and of those that are they may not yet be in Chocolatey/Scoop. And I guess, but do not know, that the version of less that ships with Git is behind all of those. Therefore the Delta README is currently pointing to @lzybkr's placeholder release https://github.com/lzybkr/less/releases/tag/fix_windows_vt

What I'm unsure about is whether these bugs affect bat also (I would think that they would, unless they affect background colors only). So I'm going to cc @sharkdp and @eth-p just in case it's helpful to them.

Some of the less issues concerned are

https://github.com/gwsw/less/issues/10 (fix merged)
https://github.com/gwsw/less/issues/42 (fix merged)
https://github.com/gwsw/less/pull/29 (unmerged PR)

**Comment by sharkdp** [comment]
> What I'm unsure about is whether these bugs affect bat also (I would think that they would, unless they affect background colors only). So I'm going to cc @sharkdp and @eth-p just in case it's helpful to them.

These issues look like they could be related to https://github.com/sharkdp/bat/issues/860 and https://github.com/sharkdp/bat/issues/887. The former has been open for some time. In the last few days, there was considerable progress thanks to the maintainer of less who fixed both bugs and thanks to @jftuga who did great work on building and providing binaries in order for Chocolatey and Scoop to update their versions of `less` (https://github.com/jftuga/less-Windows). Chocolatey has already upgraded their version of `less` and there is an open PR for Scoop: https://github.com/ScoopInstaller/Main/pull/1108

**Comment by nzbart** [comment]
The version of less that comes with Git from Chocolatey is:
>less 551 (PCRE regular expressions)
>Copyright (C) 1984-2019  Mark Nudelman

Since my development machine has evolved over several years into a unique snowflake, I tried the new package on a clean install of Windows 10 by following these steps:

1. Install Chocolatey and some packages for testing in an admin PowerShell window:
    ```
    Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iwr https://chocolatey.org/install.ps1 -UseBasicParsing | iex
    choco install -y git delta microsoft-windows-terminal
    exit
    ```

2. Start a new PowerShell instance to allow path changes to take effect, then clone this repository for testing:

    ```
    git clone https://github.com/dandavison/delta.git
    cd delta
    git config core.pager "delta --dark"
    exit
    ```

3. Run the following in Windows Terminal and stock PowerShell:
    ```
    cd delta
    git log -p 
    ```

The result is below:

![image](https://user-images.githubusercontent.com/1592130/82706163-26b75800-9ccd-11ea-88f9-eec64e9407e5.png)

Note that installing the Chocolatey `less` package didn't help because Git still uses the bundled one.

What did work is overwriting the Git `less` pager with the one that you linked to (run as admin in PowerShell):

```
iwr -Uri 'https://github.com/lzybkr/less/releases/download/fix_windows_vt/less.exe' -OutFile 'C:\Program Files\Git\usr\bin\less.exe'
```

In conclusion, it appears that:
1. There is definitely a problem with rendering out of the box with my Chocolatey package as it currently stands.
2. The fix is not as simple as depending on the `less` Chocolatey package like `bat` does.
3. There is another - unidentified - way to solve this problem, since I didn't have the issue on my dev PC and I haven't overwritten the Git less.exe pager.

I'm happy to add a note about these problems on the package, although it would be great if there was a concise section in the readme that I could link to for more detail.

**Comment by dandavison** [comment]
Extremely helpful @nzbart, thanks.

Question about this part:

> Note that installing the Chocolatey less package didn't help because Git still uses the bundled one.

I believe that the behaviour of Git to use its bundled less is not relevant here: however, perhaps the Git `less` simply has `$PATH` priority?


The reason I say Git's behaviour is not relevant is:  when you did `git log -p`, you had already done `git config core.pager "delta --dark"`. Therefore `git` was no longer starting a `less` process itself: instead it was invoking `delta`. `delta` (like `bat`) spawns a child `less` process by using Rust's `std::process`. The executable that is uses is determined by the values of the env vars `BAT_PAGER` and `PAGER`, with fall back to simply the string "less". (In `bat`'s case there is also a `--pager` command line option, but delta doesn't have that.) I don't know how `$PATH` resolution works on Windows but I would think that as long as your Chocolatey `less` has `$PATH` priority over Git's less, from the point of view of the delta process, then Delta would resolve to it. Does that sound right to you?

**Comment by nzbart** [comment]
@dandavison The Chocolatey Git package by default doesn't put tools like `less` on the path, but Git does prepend the path containing those tools at runtime when it calls a pager like delta. Because it prepends the path, the less.exe packaged with Git will take precedence over any other installation of less. 

The paths it prepends are:
* `C:\Program Files\Git\mingw64\libexec\git-core`
* `C:\Program Files\Git\mingw64\bin`
* `C:\Program Files\Git\usr\bin`
* `C:\Users\<username>\bin`

The packaged less.exe is in `C:\Program Files\Git\usr\bin`.

Note that we could deploy the modified less.exe alongside delta.exe and instruct delta to use that copy. This could be done either via a command-line argument in delta, or via modifying the path in the git config for `core.pager`.

**Comment by jftuga** [comment]
> The version of less that comes with Git from Chocolatey is:
> 
> > less 551 (PCRE regular expressions)
> > Copyright (C) 1984-2019  Mark Nudelman

@nzbart 

The Chocolatey version of less is now 562.  Could you try / test this version?

https://chocolatey.org/packages/Less

**Comment by dandavison** [comment]
> The Chocolatey Git package by default doesn't put tools like less on the path, but Git does prepend the path containing those tools at runtime when it calls a pager like delta.

Ohhh.. thanks for explaining that @nzbart! Until you said that I hadn't understood the different considerations facing bat and delta regarding `less` on Windows but that makes it clear.

> Note that we could deploy the modified less.exe alongside delta.exe and instruct delta to use that copy. This could be done either via a command-line argument in delta

Right, we could add `--pager` (like `bat`).

> via modifying the path in the git config for core.pager.

Can you spell out what this one would look like?

Would another option be for delta,  on Windows, to fiddle with the value of the path env var in the delta process, essentially pushing `*\Git\*` paths back to the end of the search path? It seems like that might yield the desired result out-of-the-box.

**Comment by nzbart** [comment]
@dandavison My thoughts are that I could package your recommended less executable alongside delta.exe and somehow get delta to use it when available. The core.pager option I mentioned above does this, but puts the onus onto the user to prepend the path to the modified less.exe in their core.pager config:
```ini
[core]
       pager = "powershell.exe -Command \"\\$env:PATH=('C:\\ProgramData\\chocolatey\\lib\\delta\\tools;' + \\$(\\$env:PATH)); delta --dark\""
```
While this technically works, requiring the user to understand the cryptic use of escaping makes this an unpalatable option.

My preferred option would be to deploy less.exe alongside delta.exe and modify delta to first look for less alongside itself before resorting to the one on the path. I like this approach because it should work in most scenarios and is unintrusive since it doesn't modify the system less.exe globally.

**Comment by dandavison** [comment]
The Delta README now instructs Windows users to use the latest `less` from https://github.com/jftuga/less-Windows/releases/latest, since I believe all relevant changes are in the version users will get that way. The env var `DELTA_PAGER` is available to specify an absolute path to the pager executable, in addition to the standard `PAGER` env var. I'll close this for now, but @nzbart do please comment further or re-open if more is needed to support installation in Windows.
